front/src/app/states/user.state.ts

```diff
@@ -0,0 +1,72 @@
+import {State, Action, StateContext, Selector} from '@ngxs/store';
+import {User} from '../models/user';
+import {UserService} from '../services/user.service';
+import {AddUser, DeleteUser, GetUsers, UpdateUser} from '../actions/user.action';
+import {tap} from 'rxjs/operators';
+
+export class UserStateModel {
+    users: User[];
+}
+
+@State<UserStateModel>({
+    name: 'users',
+    defaults: {
+        users: []
+    }
+})
+
+export class UserState {
+    constructor(private userService: UserService) { }
+
+    @Selector()
+    static getUserList(state: UserStateModel) {
+        return state.users;
+    }
+
+    @Action(GetUsers)
+    getUsers({getState, setState}: StateContext<UserStateModel>) {
+        return this.userService.fetchUsers().pipe(tap((result) => {
+            const state = getState();
+            setState({
+                ...state,
+                users: result,
+            });
+        }));
+    }
+
+    @Action(AddUser)
+    addUser({getState, patchState}: StateContext<UserStateModel>, {payload}: AddUser) {
+        return this.userService.addUser(payload).pipe(tap((result) => {
+            const state = getState();
+            patchState({
+                users: [...state.users, result]
+            });
+        }));
+    }
+
+    @Action(DeleteUser)
+    deleteUser({getState, setState}: StateContext<UserStateModel>, {id}: DeleteUser) {
+        return this.userService.deleteUser(id).pipe(tap(() => {
+            const state = getState();
+            const filteredArray = state.users.filter(item => item.id !== id);
+            setState({
+                ...state,
+                users: filteredArray
+            });
+        }));
+    }
+
+    @Action(UpdateUser)
+    updateUser({getState, setState}: StateContext<UserStateModel>, {payload, id}: UpdateUser) {
+        return this.userService.updateUser(payload, id).pipe(tap((result) => {
+            const state = getState();
+            const userList = [...state.users];
+            const userIndex = userList.findIndex(item => item.id === id);
+            userList[userIndex] = result;
+            setState({
+                ...state,
+                users: userList,
+            });
+        }));
+    }
+}
```

Comment:
Dodawanie, usuwanie itd. z palca wewnątrz akcji nie jest dobre. Skąd pewność, że backend przyjął request o np. dodanie użytkownika? Frontend pokaże, że użytkownik jest dodany, a tak się na prawdę nie stało i już jest problem. Lepiej po wywołaniu akcji, która coś zmienia , wywołać na koniec pobranie informacji z backendu i to też tylko w razie powodzenia akcji, bo jak się nie udało, to nie ma sensu zaciągać świeżych informacji, skoro i tak nic się nie zmieniło.